drivers/input/input_stmpe811.c

```diff
@@ -0,0 +1,481 @@
+		}
+
+		ret = i2c_reg_write_byte_dt(&config->bus, STMPE811_FIFO_STA_REG,
+					    FIFO_STA_OPERATIONAL);
+		if (ret < 0) {
+			return ret;
+		}
+	} else if (touch_detected == 0) {
+		data->touch_detected = 0;
+	} else {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int stmpe811_ts_enable_it(const struct device *dev)
+{
+	const struct stmpe811_config *config = dev->config;
+
+	/* Enable FIFO above threshold interrupt */
+	int err = i2c_reg_update_byte_dt(&config->bus, STMPE811_INT_EN_REG,
+					 BIT_MASK(8), STMPE811_GIT_FTH);
+
+	if (err < 0) {
+		return err;
+	}
+
+	/* Enable global interrupts */
+	return i2c_reg_update_byte_dt(&config->bus, STMPE811_INT_CTRL_REG,
+				      BIT_MASK(8), STMPE811_GIT_EN);
+}
+
+static int stmpe811_clear_global_it(const struct device *dev, uint8_t source)
+{
+	const struct stmpe811_config *config = dev->config;
+
+	/* Write 1 to the bits that have to be cleared */
+	return i2c_reg_write_byte_dt(&config->bus, STMPE811_INT_STA_REG, source);
+}
+
+static int stmpe811_process(const struct device *dev)
+{
+	const struct stmpe811_config *config = dev->config;
+	struct stmpe811_data *data = dev->data;
+
+	int err = stmpe811_ts_get_state(dev);
+
+	if (err) {
+		return err;
+	}
+
+	if (data->touch_detected) {
+		int x = config->screen_width - (data->touch_x / 4096.0) * config->screen_width;
+		int y = config->screen_height - (data->touch_y / 4096.0) * config->screen_height;
+
+		input_report_abs(dev, INPUT_ABS_X, x, false, K_FOREVER);
+		input_report_abs(dev, INPUT_ABS_Y, y, false, K_FOREVER);
+		input_report_key(dev, INPUT_BTN_TOUCH, 1, true, K_FOREVER);
+	}
+
+	/* Start polling to detect release */
+	data->poll_release_counter = 0;
+	k_work_schedule(&data->polling_work, K_MSEC(config->polling_period_ms));
+
+	return 0;
+}
+
+static void stmpe811_work_handler(struct k_work *work)
+{
+	struct stmpe811_data *data = CONTAINER_OF(work, struct stmpe811_data, work);
+
+	stmpe811_process(data->dev);
+}
+
+static void stmpe811_poll_release_work_handler(struct k_work *work)
+{
+	struct k_work_delayable *polling_work = CONTAINER_OF(work, struct k_work_delayable, work);
+	struct stmpe811_data *data = CONTAINER_OF(polling_work, struct stmpe811_data, polling_work);
+
+	const struct device *dev = data->dev;
+	const struct stmpe811_config *config = dev->config;
+
+	bool poll_again = true;
+	int err;
+	uint32_t current_cycles;
+	uint32_t cycles_diff;
+	uint32_t wait_period_us;
+
+	k_timepoint_t poll_time_end = sys_timepoint_calc(K_MSEC(config->hold_timeout_ms));
+	uint32_t start_period_cycles = k_cycle_get_32();
+	bool touch = stmpe811_ts_detect_touch(dev);
+
+	/* Clear FIFO - data->poll_period_us has to be high enough
+	 * to fill it up before next
+	 */
+
+	err = i2c_reg_write_byte_dt(&config->bus, STMPE811_FIFO_STA_REG, FIFO_STA_CLEAR);
+	err += i2c_reg_write_byte_dt(&config->bus, STMPE811_FIFO_STA_REG,
+						FIFO_STA_OPERATIONAL);
+
+	if (err < 0) {
+		LOG_ERR("Could not clear FIFO after polling (%d)", err);
+	} else {
```

Comment:
just do the error check after each call end return on error, then the normal execution path (the block below) gets indented back